scripts/ci/test_plan.py

```diff
@@ -4,23 +4,133 @@
+                continue
+            if re.match('^samples', item):
+                _test_apps_path.append(item)
+                continue
+            if os.path.isfile(os.path.join(zephyr_base_dir, item)):
+                __f = os.path.basename(item)
+                if __f in target_files:
+                    matched = True
+            else:
+                if '*' in item:
+                    _files = Path(zephyr_base_dir).rglob(item)
+                    _files = list(set(_files))
+                    __f = [os.path.basename(i) for i in _files]
+                else:
+                    _files = Path(os.path.join(zephyr_base_dir, item)).rglob("*")
+                    _files = list(set(_files))
+                    __f = [os.path.basename(i) for i in _files]
+
+                if [ e for e in __f if e in target_files ]:
+                    matched = True
+
+        if not matched:
+            _test_apps_path.clear()
+        else:
+            test_apps_path += _test_apps_path
+
+
+    return test_apps_path
+
+def _match_tests_with_file(zephyr_base_dir, target_files):
+    ''' match_tests_with_file
+        match related applications with given modified files
+        based on MAINTAINERS.yml
+
+        :param zephyr_base_dir: zephyr base directory
+        :type zephyr_base_dir: string
+
+        :param target_files: target files without path
+        :type target_files: array
+    '''
+    test_apps_path = []
+    filtered_target_files = []
+
+    for file in target_files:
+        extension = os.path.splitext(file)[1]
+        if extension in ['.c', '.h']:
+            filtered_target_files.append(file)
+
+    for path in Path(zephyr_base_dir).rglob('MAINTAINERS.yml'):
+        # only porocess the first found
+        with open(path, "r", encoding='UTF-8') as stream:
+            try:
+                data_loaded = yaml.safe_load(stream)
+                break
+            except yaml.YAMLError as exc:
+                logging.debug(exc)
+                return set()
+
+    items = data_loaded.items()
+
+    with concurrent.futures.ProcessPoolExecutor(max_workers=os.cpu_count()) as executor:
+        futures = [executor.submit(_procee_maintainer_file, item,
+                    zephyr_base_dir, filtered_target_files) for _k, item in items]
+        for future in concurrent.futures.as_completed(futures):
+            test_apps_path += future.result()
+
+    logging.info(test_apps_path)
+    # find all applications in test_apps_path
+    test_path_list = []
+    for _tp in test_apps_path:
+        _apps_path = os.path.join(zephyr_base_dir, _tp)
+        for path in Path(_apps_path).rglob("testcase.yaml"):
+            _tp = os.path.relpath(path, zephyr_base_dir)
+            test_path_list.append(os.path.dirname(_tp))
+        for path in Path(_apps_path).rglob("sample.yaml"):
+            _tp = os.path.relpath(path, zephyr_base_dir)
+            test_path_list.append(os.path.dirname(_tp))
+
```

Comment:
Can you refactor this part, so the same code is not repeated?

Reply:
ok